### ISSUE TITLE
no-unused-vars considers destructured args to be args instead of local vars

### ISSUE THREAD

**Issue by exogen** [eslint/eslint#11879]
**What rule do you want to change?**

When `no-unused-vars` has its `args` option set to `none`, it does not warn about this situation:

```js
// Exhibit A
function foo({ a, b, c }) {
  return a;
}
```

In this case it is considering `a`, `b`, and `c` to be args. But, they are not function args – they are properties on the single object argument to the function. (Pedantic side note: the single unnamed argument will always be considered "used" by the `no-unused-vars` rule by virtue of being destructured – there is no way *not* to use it).

The current behavior is surprising to me, [and perhaps others as well](https://github.com/eslint/eslint/issues/11712), because destructuring an arg should be equivalent to:

```js
// Exhibit B
function foo(options) {
  let { a, b, c } = options;
  return a;
}
```

Which, even with `args` set to `none`, does produce the expected warning:

```
error  'b' is assigned a value but never used  no-unused-vars
error  'c' is assigned a value but never used  no-unused-vars
```

I propose either making the `args` option not apply to destructured properties of actual args, or adding a new `args` option that does not conflate arguments themselves with properties that are destructured from those arguments.

What I really do not want to do is tell my team they should start writing their code like in Exhibit B instead of Exhibit A just because of a quirk in what a particular ESLint rule considers to be an "arg" – we shouldn't have to change our code to get the desired behavior. :)

FWIW, the current behavior has resulted in an actual bug for me.

**Does this change cause the rule to produce more or fewer warnings?**

More.

**How will the change be implemented? (New option, new default behavior, etc.)?**

Either a change in what is considered an `arg` (only the function's actual top-level params would be consider args – the things you call the function with, and the things counted by `fn.length`).

OR a new option for `no-unused-vars` that indicates whether you'd like to treat the destructured properties of unnamed arguments as arguments themselves or not for the purpose of this rule.

**Please provide some example code that this change will affect:**

```js
function foo({ a, b, c }) {
  return a;
}
```

**What does the rule currently do for this code?**

With `args: none`, there is no error.

**What will the rule do after it's changed?**

With `args: none`, it should warn:

```
error  'b' is assigned a value but never used  no-unused-vars
error  'c' is assigned a value but never used  no-unused-vars
```

Because a, b, and c are *not* args, but rather properties of an arg.

**Are you willing to submit a pull request to implement this change?**

If requested, but it would take a while.
